node/src/exchange/server.rs

```diff
@@ -1,52 +1,396 @@
+            store,
+            store_jobs: FuturesUnordered::new(),
+        }
     }
 
-    #[instrument(level = "trace", skip(self, _respond_to))]
-    pub(super) fn on_request_received(
+    #[instrument(level = "trace", skip(self, response_channel))]
+    pub(super) fn on_request_received<ID>(
         &mut self,
         peer: PeerId,
-        request_id: RequestId,
+        request_id: ID,
         request: HeaderRequest,
-        _respond_to: ResponseChannel<Vec<HeaderResponse>>,
-    ) {
-        // TODO
+        response_channel: R::Channel,
+    ) where
+        ID: Display + Debug,
+    {
+        let Some((amount, data)) = self.parse_request(request) else {
+            self.store_jobs
+                .push(handle_invalid_request(response_channel).boxed());
+            return;
+        };
+
+        let store_job = match data {
+            header_request::Data::Origin(0) => {
+                handle_request_current_head(self.store.clone(), response_channel).boxed()
+            }
+            header_request::Data::Origin(height) => {
+                handle_request_by_height(self.store.clone(), response_channel, height, amount)
+                    .boxed()
+            }
+            header_request::Data::Hash(hash) => {
+                handle_request_by_hash(self.store.clone(), response_channel, hash).boxed()
+            }
+        };
+
+        self.store_jobs.push(store_job);
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_response_sent(&mut self, peer: PeerId, request_id: RequestId) {
-        // TODO
+        trace!("response_sent; request_id: {request_id}, peer: {peer}");
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_failure(
         &mut self,
         peer: PeerId,
         request_id: RequestId,
         error: InboundFailure,
     ) {
-        // TODO
+        // TODO: cancel job if libp2p already failed it?
+        trace!("on_failure; request_id: {request_id}, peer: {peer}, error: {error:?}");
+    }
+
+    fn parse_request(&self, request: HeaderRequest) -> Option<(u64, header_request::Data)> {
+        if !request.is_valid() {
+            return None;
+        }
+
+        let HeaderRequest {
+            amount,
+            data: Some(data),
+        } = request
+        else {
+            return None;
+        };
+
+        Some((amount, data))
+    }
+
+    pub fn poll(&mut self, cx: &mut Context<'_>, sender: &mut R) -> Poll<()> {
+        while let Poll::Ready(Some((channel, response))) =
+            Pin::new(&mut self.store_jobs).poll_next(cx)
+        {
+            sender.send_response(channel, response);
+        }
+
+        Poll::Pending
+    }
+}
+
+async fn handle_request_current_head<S, C>(store: Arc<S>, channel: C) -> (C, ResponseType)
+where
+    S: Store,
+{
+    let response = match store.get_head().await {
+        Ok(head) => head.to_header_response(),
+        Err(_) => HeaderResponse::not_found(),
+    };
```

Comment:
This is a good candidate for `unwrap_or_else`:

```rust
            let response = store
                .get_head()
                .await
                .map(|head| head.to_header_response())
                .unwrap_or_else(|_| HeaderResponse::not_found());
```

Reply:
switche to functional style 👍